tests/unit/lms/views/api/canvas/authorize_test.py

```diff
@@ -110,6 +102,66 @@ def test_it_passes_authorize_url_to_the_template(
             == "http://example.com/api/canvas/authorize?authorization=TEST_AUTHORIZATION_PARAM"
         )
 
+    def test_it_doesnt_pass_authorize_url_if_theres_no_authorized_user(
+        self, pyramid_request, views
+    ):
+        pyramid_request.lti_user = None
+
+        template_variables = views.oauth2_redirect_error()
+
+        assert "authorize_url" not in template_variables
+
+    @pytest.mark.parametrize(
+        "params,invalid_scope",
+        [
+            ({"error": "invalid_scope"}, True),
+            ({"error": "unknown_error"}, False),
+            ({"foo": "bar"}, False),
+        ],
+    )
+    def test_it_tells_the_template_whether_to_show_the_missing_scopes_error_message(
+        self, pyramid_request, params, invalid_scope, views
+    ):
+        pyramid_request.params.clear()
+        pyramid_request.params.update(params)
+
+        template_variables = views.oauth2_redirect_error()
+
+        assert template_variables["invalid_scope"] == invalid_scope
+
+    @pytest.mark.parametrize(
+        "params,expected_details",
+        [
+            (
+                {"error_description": mock.sentinel.error_description},
+                mock.sentinel.error_description,
+            ),
+            ({"foo": "bar"}, None),
+        ],
+    )
+    def test_it_passes_error_descriptions_from_Canvas_to_the_template(
+        self, pyramid_request, params, expected_details, views
+    ):
+        pyramid_request.params.clear()
+        pyramid_request.params.update(params)
+
+        template_variables = views.oauth2_redirect_error()
+
+        assert template_variables["details"] == expected_details
+
+    def test_it_passes_our_required_API_scopes_to_the_template(self, views):
+        template_variables = views.oauth2_redirect_error()
+
+        assert template_variables["scopes"] == (
+            "url:GET|/api/v1/courses/:course_id/files",
+            "url:GET|/api/v1/files/:id/public_url",
+        )
```

Comment:
```suggestion
        assert template_variables["scopes"] == views.scopes
```

Do we want to say the exact scopes, or just that we pass through "the" scopes? It means not changing this in future if we do, but it's less specific. I lean towards just asserting they are the same, otherwise our tests tend towards an MD5 of our code.